### DISCUSSION TITLE
Why does ImmutableList throw IndexOutOfRangeException instead of ArgumentOutOfRangeException when getting the element of the set at the given index?

### DISCUSSION THREAD

**Discussion by PonchoPowers** [dotnet/runtime#108366] [Q&A] [answered]
Why does ImmutableList throw IndexOutOfRangeException instead of ArgumentOutOfRangeException when getting the element of the set at the given index, but List throws an ArgumentOutOfRangeException exception?

ImmutableList

```
/// <exception cref="IndexOutOfRangeException">Thrown from getter when <paramref name="index"/> is negative or not less than <see cref="Count"/>.</exception>
        public T this[int index] => _root.ItemRef(index);
```

List
```
        public T this[int index] {
            get {
                // Following trick can reduce the range check by one
                if ((uint) index >= (uint)_size) {
                    ThrowHelper.ThrowArgumentOutOfRangeException();
                }
                Contract.EndContractBlock();
                return _items[index]; 
            }
```

**Comment by dotnet-policy-service** [comment]
Tagging subscribers to this area: @dotnet/area-system-collections
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.

**Comment by huoyaoyuan** [comment]
It's probably for consistency with `ImmutableArray`, which must depends on `IndexOutOfRangeException` on array access for performance reason.

**Comment by PonchoPowers** [comment]
I can't understand how using `IndexOutOfRangeException` over `ArgumentOutOfRangeException` can be any more performant?

I spent more time looking at this and now think `List` is throwing the wrong exception and `ImmutableList` is actually throwing the correct exception, because any range exceptions with indexers are supposed to throw an `IndexOutOfRangeException`  instead of an `ArgumentOutOfRangeException` according to the docs.

**Reply by huoyaoyuan** [reply]
`ImmutableArray` is a slim wrapper of array. To be performant, it must act as an array without anything extra and array uses `IndexOutOfRangeException`.

**Comment by AustinWise** [comment]
It appears the [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablelist-1.item?view=net-8.0) does not agree with the [code](https://www.nuget.org/packages/System.Collections.Immutable/8.0.0). That is, the docs say `IndexOutOfRangeException` is thrown but the code throws a `ArgumentOutOfRangeException`. When I run this small program:

```csharp
using System.Collections.Immutable;

ImmutableList<int> list = ImmutableList.Create<int>([]);
Console.WriteLine(list[0]);
```

It prints:

```
Unhandled exception. System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'index')
   at System.Collections.Immutable.Requires.FailRange(String parameterName, String message)
   at System.Collections.Immutable.Requires.Range(Boolean condition, String parameterName, String message)
   at System.Collections.Immutable.ImmutableList`1.Node.ItemRef(Int32 index)
   at System.Collections.Immutable.ImmutableList`1.get_Item(Int32 index)
   at Program.<Main>$(String[] args)
```

You can see the source for `FailRange` [here](https://github.com/dotnet/runtime/blob/release/8.0/src/libraries/System.Collections.Immutable/src/Validation/Requires.cs#L97).

[Godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:csharp,selection:(endColumn:36,endLineNumber:11,positionColumn:36,positionLineNumber:11,selectionStartColumn:36,selectionStartLineNumber:11,startColumn:36,startLineNumber:11),source:'%23nullable+disable%0A%0Ausing+System%3B%0Ausing+System.Collections.Immutable%3B%0A%0Aclass+Program%0A%7B%0A++++static+void+Main()%0A++++%7B%0A++++++++ImmutableList%3Cint%3E+list+%3D+ImmutableList.Create%3Cint%3E(%5B%5D)%3B%0A++++++++Console.WriteLine(list%5B0%5D)%3B%0A++++%7D%0A%7D%0A'),l:'5',n:'1',o:'C%23+source+%231',t:'0')),k:33.86931715084305,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:dotnettrunkcsharpcoreclr,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:3,lang:csharp,libs:!(),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+.NET+CoreCLR+(main)+(Editor+%231)',t:'0')),k:32.79734951582363,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+clang+(trunk)',editorid:1,fontScale:14,fontUsePx:'0',j:3,wrap:'1'),l:'5',n:'0',o:'Output+of+.NET+CoreCLR+(main)+(Compiler+%233)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Reply by AustinWise** [reply, marked as answer]
This will be fixed by #108896 and dotnet/dotnet-api-docs#10573.
